cli/deno_dir.rs

```diff
@@ -534,100 +712,95 @@ fn filter_shebang(bytes: Vec<u8>) -> Vec<u8> {
-        media_type: map_content_type(&p, Some(&content_type)),
-        source_code: source.as_bytes().to_owned(),
-        maybe_output_code_filename: None,
-        maybe_output_code: None,
-        maybe_source_map_filename: None,
-        maybe_source_map: None,
-      }))
-    },
-  )
+#[derive(Debug)]
+pub struct SourceMetaData {
+  // MIME type of the source code.
+  pub mime_type: Option<String>,
+  // Where should we actually look for source code.
+  // This should be an absolute path with protocol!
+  pub redirect_source: Option<String>,
 }
 
-// Prototype https://github.com/denoland/deno/blob/golang/deno_dir.go#L37-L73
-/// Fetch remote source code.
-#[cfg(test)]
-fn fetch_remote_source(
-  module_name: &str,
-  filename: &str,
-) -> DenoResult<Option<ModuleMetaData>> {
-  tokio_util::block_on(fetch_remote_source_async(module_name, filename))
+static MIME_TYPE: &'static str = "mime_type";
+static REDIRECT_SOURCE: &'static str = "redirect_source";
+
+fn source_metadata_filename(filename: &str) -> String {
+  [&filename, ".meta"].concat()
+}
+
+fn get_source_metadata(filename: &str) -> SourceMetaData {
+  let metadata_filename = source_metadata_filename(filename);
+  let md = Path::new(&metadata_filename);
+  // .meta file might not exists.
+  // This is okay for local source.
+  let maybe_metadata_string = fs::read_to_string(&md).ok();
+  if let Some(metadata_string) = maybe_metadata_string {
+    // TODO(kevinkassimo): consider introduce serde::Deserialize to make things simpler.
+    let maybe_metadata: serde_json::Result<serde_json::Value> =
+      serde_json::from_str(&metadata_string);
+    if let Ok(metadata) = maybe_metadata {
+      return SourceMetaData {
+        mime_type: metadata[MIME_TYPE].as_str().map(String::from),
+        redirect_source: metadata[REDIRECT_SOURCE].as_str().map(String::from),
+      };
+    }
+  }
+  SourceMetaData {
+    mime_type: None,
+    redirect_source: None,
+  }
 }
 
-/// Fetch local or cached source code.
-fn fetch_local_source(
-  module_name: &str,
+/// Save metadata related to source filename to {filename}.meta file,
+/// only when there is actually something necessary to save.
+fn save_source_metadata(
   filename: &str,
-) -> DenoResult<Option<ModuleMetaData>> {
-  let p = Path::new(&filename);
-  let media_type_filename = [&filename, ".mime"].concat();
-  let mt = Path::new(&media_type_filename);
-  let source_code = match fs::read(p) {
-    Err(e) => {
-      if e.kind() == std::io::ErrorKind::NotFound {
-        return Ok(None);
-      } else {
-        return Err(e.into());
-      }
+  mime_type: Option<String>,
+  redirect_source: Option<String>,
+) {
+  let metadata_filename = source_metadata_filename(filename);
+  // Remove possibly existing stale .meta file.
+  // May not exist. DON'T unwrap.
+  let _ = std::fs::remove_file(&metadata_filename);
+  let p = PathBuf::from(filename);
+  // TODO(kevinkassimo): consider introduce serde::Deserialize to make things simpler.
```

Comment:
Yes, we need to do this soon. There's another place in the source map handler where I'm manually parsing JSON like this.